nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,294 @@
+                        path = mkOption {
+                          type = nullOr str;
+                          description = "Remote path for the repository.";
+                        };
+                      };
+                    }
+                  )
+                );
+                default = [ ];
+                description = "List of repositories to backup to.";
+              };
+
+              tags = mkOption {
+                type = attrsOf str;
+                default = {
+                  inherit name;
+                };
+                description = "List of tags for this job.";
+              };
+
+              startAt = mkOption {
+                type = either str (listOf str);
+                default = [ ];
+                description = ''
+                  Automatically start this backup at the given date/time, which
+                  must be in the format described in {manpage}`systemd.time(7)`.
+                '';
+              };
+
+              extraOptions = mkOption {
+                type = listOf str;
+                default = [ ];
+                description = ''
+                  Additional options passed to `bupstash put`.
+                '';
+              };
+            };
+
+            config = {
+              extraOptions = builtins.map (pattern: "--exclude '${pattern}'") config.excludes;
+            };
+          }
+        )
+      );
+      default = { };
+      description = "List of backup jobs to run using bupstash.";
+    };
+  };
+
+  config = mkMerge [
+    (mkIf (cfg.jobs != {}) {
+      systemd.services =
+        mapAttrs'
+          (
+            name: job:
+            nameValuePair "bupstash-${name}" (
+              let
+                keysDir = "keys/${name}";
+                allRepositories = builtins.concatMap (job: job.repositories) (builtins.attrValues cfg.jobs);
+                hasSSH = builtins.any builtins.isAttrs allRepositories;
+              in
+              {
+                inherit (job) startAt;
+
+                path = [ pkgs.coreutils ] ++ (optional hasSSH pkgs.openssh);
+
+                preStart = ''
+                  mkdir -p send-logs/${name}
+
+                  if [ ! -d '${keysDir}' ]; then
+                    mkdir -p ${keysDir}
+
+                    # Generate new keys
+                    ${bupstash} new-key -o ${keysDir}/master.key
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/put.key \
+                      --put
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/list.key \
+                      --list-contents
+
+                  elif [ -f '${keysDir}/master.key' ]; then
+                    echo "[!] The master key for the job '${name}' is present, you might want to store it offline and delete it on the server."
+                  fi
```

Comment:
I'm strongly opposed to baking in key management, for at least two reasons:
- Secret keys shouldn't be moved across machine boundaries, not even temporarily
- The way you set this up makes it convergent, not declarative

Reply:
It is very reasonable, I removed this mechanism, and added a `key` option to the jobs